### PR TITLE
In-memory write caching

### PR DESCRIPTION
**Pull request by samholmes**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -131,49 +142,54 @@ export function makeMemlet(disklet: Disklet): Memlet {
...
+       * Write-through policy: write to memory cache first, then let the data be
+       * drained to disklet later.
        */
-      const dataString = JSON.stringify(data)
-
-      await disklet.setText(path, dataString)
-
       const key = getCacheKey(path)
       const file = getStoreFile(key)
+      const dataString = JSON.stringify(data)
```
Comment (samholmes) [inline src/index.ts:150]: Consider fast-json-stringify for optimization

File: src/index.ts [hunk 2]
```diff
@@ -185,50 +201,102 @@ export function makeMemlet(disklet: Disklet): Memlet {
...
+          if (state.memoryOnlyFileQueue.list().length > 0) {
+            drainMemoryOnlyFiles()
+          }
+        })
+        .catch(err => {
+          // Uh oh spaghettios
+          throw err
+        })
+    }, DRAIN_INTERVAL)
+  }
```
Comment (samholmes) [inline src/index.ts:299]: Store global `timeoutId`. Only start `setTimeout` if `timeoutId == null`. Set `timeoutId = null` if `memoryOnlyFileQueue` is empty or on caught error.

File: src/index.ts [hunk 3]
```diff
@@ -16,9 +16,17 @@ const state: MemletState = {
     memoryUsage: 0,
     files: {}
   },
-  fileQueue: makeFileQueue()
+  writtenFileQueue: makeFileQueue(),
+  memoryOnlyFileQueue: makeFileQueue()
```
[the remark this answers is not in the document]
Comment (samholmes) [reply src/index.ts:20]: Perhaps, generalize `FileQueue` to `Queue` which works on a generic type `T extends {lastTouchedTimestamp: number, key: string}`

File: src/index.ts [hunk 4]
```diff
@@ -56,6 +56,8 @@ export function makeMemlet(disklet: Disklet): Memlet {
    */
   const memletInstanceId = countOfMemletInstances++
 
+  let _nextFlushEvent: Promise<void> | undefined
```
Comment (thehobbit85) [inline src/index.ts:59]: no reason to use the underscore convention since you don't even expose this param.

File: src/index.ts [hunk 5]
```diff
@@ -296,8 +304,8 @@ export function makeMemlet(disklet: Disklet): Memlet {
    */
   function startFlushing(): Promise<void> {
     // If timeout is already running then do nothing
-    if (memlet._nextFlushEvent != null) {
-      return memlet._nextFlushEvent
+    if (_nextFlushEvent != null) {
```
Comment (thehobbit85) [inline src/index.ts:307]: Can you please "one liner" it? ;)

File: src/queue.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
+export function makeQueue<T extends QueueItem>(): Queue<T> {
+  const items: T[] = []
+  const timestampMap: Map<string, number> = new Map()
```
Comment (thehobbit85) [inline src/queue.ts:3]: Unfortunately we can't use `Map` in any of the plugins or the the libs that are being used them.
It has some unpredictable bugs after being transpiled in the App.

**Review by thehobbit85 — CHANGES_REQUESTED** [review]
See comments.

**Review by thehobbit85 — APPROVED** [review]

**Pull request merged**